### PR TITLE
#198 Added sms and template entities and ncessary infrastructure

### LINKED ISSUE #198: Create SMS and Email template infrastructure
### Description
Create SMS and Email template with necessary infrastructure

### Requirements
- project should be runnable
- `NotificationsController` must be tested and Postman collection must be created

### Deliverables
- `NotificationType`, `NotificationTemplateType` enums
- `SmsTemplate` , `EmailTemplate` entites and DTO models
- `NotificationDbContext` DbContext
- `SmsTemplateConfiguration`, `EmailTemplateConfiguration`, `NotificationTemplateConfiguration` entity configuration
- `SmsTemplateRepository`, `EmailTemplateRepository` repositories
- `EmailTemplateService`, `SmsTemplateService` foundation service
- validation of `SmsTemplate`, `EmailTemplate`
- mapping of `SmsTemplate`, `EmailTemplate`
- `NotificationsController` controller with actions
- `GetSmsTemplates` - returns SMS templates with pagination
- `GetEmailTemplates` - returns Email templates with pagination
- `GetTemplateByType` - notification type, template type
- add migration
- register all necessary infrastructure to DI

### PR DESCRIPTION
**Pull request by Abdurahmon0412**
Closes #198

### CODE REVIEW

**Review by botirjonabdumannonovv — APPROVED** [review]

**Review by botirjonabdumannonovv — DISMISSED** [review]

File: src/Training.AirBnb.Clone.Backend/global.json [hunk 1]
```diff
@@ -0,0 +1,7 @@
+{
```
[the remark this answers is not in the document]
Comment (Abdurahmon0412) [reply src/Training.AirBnb.Clone.Backend/global.json:1]: qayerlarida qolib ketganini hammasini yozib ketsayiz yaxshi bo'lardi ancha joyiga qo'shdim docs ni

File: src/Training.AirBnb.Clone.Backend/AirBnB.Application/Common/Notifications/Services/ISmsTemplateService.cs [hunk 2]
```diff
@@ -0,0 +1,44 @@
+﻿using AirBnB.Application.Common.Models.Querying;
+using AirBnB.Domain.Entities;
+using AirBnB.Domain.Enums;
+
+namespace AirBnB.Application.Common.Notifications.Services;
+
+public interface ISmsTemplateService
+{
+    /// <summary>
+    /// it takes smsTemplate by filtering
```
Comment (sulton-max) [inline src/Training.AirBnb.Clone.Backend/AirBnB.Application/Common/Notifications/Services/ISmsTemplateService.cs:10]: Change comment to "Retrieves SMS templates by filter"

File: src/Training.AirBnb.Clone.Backend/AirBnB.Application/Common/Notifications/Services/ISmsTemplateService.cs [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+    /// <param name="asNoTracking"></param>
+    /// <param name="cancellationToken"></param>
+    /// <returns>Ilist SmsTemplate</returns>
+    ValueTask<IList<SmsTemplate>> GetByFilterAsync(
+        FilterPagination pagination,
+        bool asNoTracking = false,
+        CancellationToken cancellationToken = default);
+
+    /// <summary>
+    /// it takes smsTemplate By type
```
Comment (sulton-max) [inline src/Training.AirBnb.Clone.Backend/AirBnB.Application/Common/Notifications/Services/ISmsTemplateService.cs:22]: Change comment to "Retrieves SMS template by type"

File: src/Training.AirBnb.Clone.Backend/AirBnB.Application/Common/Notifications/Services/ISmsTemplateService.cs [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+    /// <param name="asNoTracking"></param>
+    /// <param name="cancellationToken"></param>
+    /// <returns>default or smsTemplate</returns>
+    ValueTask<SmsTemplate?> GetByTypeAsync(
+        NotificationTemplateType templateType,
+        bool asNoTracking = false,
+        CancellationToken cancellationToken = default);
+
+    /// <summary>
+    /// it creates by emailTemplate
```
Comment (sulton-max) [inline src/Training.AirBnb.Clone.Backend/AirBnB.Application/Common/Notifications/Services/ISmsTemplateService.cs:34]: Change comment to "Creates SMS template"

File: src/Training.AirBnb.Clone.Backend/AirBnB.Application/Common/Notifications/Services/IEmailTemplateService.cs [hunk 5]
```diff
@@ -0,0 +1,44 @@
+﻿using AirBnB.Application.Common.Models.Querying;
+using AirBnB.Domain.Entities;
+using AirBnB.Domain.Enums;
+
+namespace AirBnB.Application.Common.Notifications.Services;
+
+public interface IEmailTemplateService
+{
+    /// <summary>
+    /// it takes emailTemplate by filtering
```
Comment (sulton-max) [inline src/Training.AirBnb.Clone.Backend/AirBnB.Application/Common/Notifications/Services/IEmailTemplateService.cs:10]: Change just like in the comments of SMS template service

File: src/Training.AirBnb.Clone.Backend/AirBnB.Infrastructure/AirBnB.Infrastructure.csproj [hunk 6]
```diff
@@ -6,6 +6,10 @@
     <Nullable>enable</Nullable>
   </PropertyGroup>
 
+  <ItemGroup>
+    <PackageReference Include="FluentValidation.DependencyInjectionExtensions" Version="11.8.1" />
```
Comment (sulton-max) [inline src/Training.AirBnb.Clone.Backend/AirBnB.Infrastructure/AirBnB.Infrastructure.csproj:10]: Why we need dependency injection here ? Can't we just use `FluentValidation` package ?

File: src/Training.AirBnb.Clone.Backend/AirBnB.Infrastructure/Common/Validators/EmailTemplateValidator .cs [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+namespace AirBnB.Infrastructure.Common.Validators;
+
+public class EmailTemplateValidator : AbstractValidator<EmailTemplate> 
+{
+    public EmailTemplateValidator()
+    {
+        RuleFor(template => template.Content)
+            .NotEmpty()
+            .MinimumLength(10)
+            .MaximumLength(256);
```
Comment (sulton-max) [inline src/Training.AirBnb.Clone.Backend/AirBnB.Infrastructure/Common/Validators/EmailTemplateValidator .cs:14]: Change max length to 129_536

File: src/Training.AirBnb.Clone.Backend/AirBnB.Infrastructure/Common/Validators/SmsTemplateValidator.cs [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+
+namespace AirBnB.Infrastructure.Common.Validators;
+
+public class SmsTemplateValidator : AbstractValidator<SmsTemplate>
+{
+    public SmsTemplateValidator()
+    {
+        RuleFor(template => template.Content)
+            .NotEmpty()
+            .WithMessage("Sms template content is required")
```
Comment (sulton-max) [inline src/Training.AirBnb.Clone.Backend/AirBnB.Infrastructure/Common/Validators/SmsTemplateValidator.cs:13]: I think we don't need custom messages

File: src/Training.AirBnb.Clone.Backend/AirBnB.Persistence/Repositories/Interfaces/IEmailTemplateRepository.cs [hunk 9]
```diff
@@ -0,0 +1,27 @@
+﻿using AirBnB.Domain.Entities;
+using System.Linq.Expressions;
+
+namespace AirBnB.Persistence.Repositories.Interfaces;
+
+public interface IEmailTemplateRepository
+{
+    /// <summary>
+    /// Get all templates which IQuerable
```
Comment (sulton-max) [inline src/Training.AirBnb.Clone.Backend/AirBnB.Persistence/Repositories/Interfaces/IEmailTemplateRepository.cs:9]: Change comment to "Returns queryable of email templates"

File: src/Training.AirBnb.Clone.Backend/AirBnB.Persistence/Repositories/Interfaces/IEmailTemplateRepository.cs [hunk 10]
```diff
@@ -0,0 +1,27 @@
...
+    /// </summary>
+    /// <param name="predicate"></param>
+    /// <param name="asNoTracking"></param>
+    /// <returns>IQueryable<EmailTemplate> Get </returns>
+    IQueryable<EmailTemplate> Get(
+        Expression<Func<EmailTemplate, bool>>? predicate = default,
+        bool asNoTracking = false);
+
+    /// <summary>
+    /// Create EmailTemplate async
```
Comment (sulton-max) [inline src/Training.AirBnb.Clone.Backend/AirBnB.Persistence/Repositories/Interfaces/IEmailTemplateRepository.cs:19]: Don't use "async" keyword

File: src/Training.AirBnb.Clone.Backend/AirBnB.Persistence/Repositories/Interfaces/ISmsTemplateRepository.cs [hunk 11]
```diff
@@ -0,0 +1,28 @@
...
+    /// <summary>
+    /// Create SmsTemplate async
+    /// </summary>
+    /// <param name="predicate"></param>
+    /// <param name="saveChanges"></param>
+    /// <returns>ValueTask<SmsTemplate</returns>
+    ValueTask<SmsTemplate> CreateAsync(
+        SmsTemplate smsTemplate,
+        bool saveChanges = true,
+        CancellationToken cancellationToken = default);
```
Comment (sulton-max) [inline src/Training.AirBnb.Clone.Backend/AirBnB.Persistence/Repositories/Interfaces/ISmsTemplateRepository.cs:27]: Update comments just like in email template repository comments

File: src/Training.AirBnb.Clone.Backend/AirBnB.Application/Common/Models/Querying/FilterPagination.cs [hunk 12]
```diff
@@ -0,0 +1,11 @@
+﻿namespace AirBnB.Application.Common.Models.Querying;
+
+/// <summary>
+/// Created pagination for templates
+/// </summary>
+public class FilterPagination
+{
+    public uint PageSize { get; set; } = 10;
+
+    public uint PageToken { get; set; } = 1;
```
Comment (sulton-max) [inline src/Training.AirBnb.Clone.Backend/AirBnB.Application/Common/Models/Querying/FilterPagination.cs:10]: Add comments

File: src/Training.AirBnb.Clone.Backend/AirBnB.Api/appsettings.json [hunk 13]
```diff
@@ -5,5 +5,6 @@
       "Microsoft.AspNetCore": "Warning"
     }
   },
-  "AllowedHosts": "*"
-}
+  "AllowedHosts": "*",
+  "AirBnBNotificationsDatabaseConnection": "Host=localhost;Port=5433;Database=TrainningAirBnB;Username=postgres;Password=postgres"
```
Comment (sulton-max) [inline src/Training.AirBnb.Clone.Backend/AirBnB.Api/appsettings.json:9]: Add connection strings inside "ConnectionStrings" section and move connection strings to appsettings.Development.json

File: src/Training.AirBnb.Clone.Backend/AirBnB.Api/Profiles/SmsTemplateProfile.cs [hunk 14]
```diff
@@ -0,0 +1,19 @@
+﻿using AirBnB.Api.Models.DTOs;
+using AirBnB.Domain.Entities;
+using AutoMapper;
+
+namespace AirBnB.Api.Profiles;
+
+/// <summary>
+/// AutoMapper profile for mapping between the SmsTemplate and SmsTemplateDto classes.
+/// </summary>
+public class SmsTemplateProfile : Profile
```
Comment (sulton-max) [inline src/Training.AirBnb.Clone.Backend/AirBnB.Api/Profiles/SmsTemplateProfile.cs:10]: Rename to "SmsTemplateMapper", both type and the file

File: src/Training.AirBnb.Clone.Backend/AirBnB.Api/Profiles/EmailTemplateProfile.cs [hunk 15]
```diff
@@ -0,0 +1,19 @@
+﻿using AirBnB.Api.Models.DTOs;
+using AirBnB.Domain.Entities;
+using AutoMapper;
+
+namespace AirBnB.Api.Profiles;
+
+/// <summary>
+/// AutoMapper profile for mapping between the EmailTemplate and EmailTemplateDto classes.
+/// </summary>
+public class EmailTemplateProfile : Profile
```
Comment (sulton-max) [inline src/Training.AirBnb.Clone.Backend/AirBnB.Api/Profiles/EmailTemplateProfile.cs:10]: Rename to "EmailTemplateMapper", both type and the file

File: src/Training.AirBnb.Clone.Backend/AirBnB.Api/Models/DTOs/SmsTemplateDTO.cs [hunk 16]
```diff
@@ -0,0 +1,14 @@
+﻿using AirBnB.Domain.Enums;
+
+namespace AirBnB.Api.Models.DTOs;
+
+public class SmsTemplateDTO
```
Comment (sulton-max) [inline src/Training.AirBnb.Clone.Backend/AirBnB.Api/Models/DTOs/SmsTemplateDTO.cs:5]: "Dto" as postfix, not "DTO"

File: src/Training.AirBnb.Clone.Backend/AirBnB.Api/Models/DTOs/EmailTemplateDTO.cs [hunk 17]
```diff
@@ -0,0 +1,16 @@
+﻿using AirBnB.Domain.Enums;
+
+namespace AirBnB.Api.Models.DTOs;
+
+public class EmailTemplateDTO
```
Comment (sulton-max) [inline src/Training.AirBnb.Clone.Backend/AirBnB.Api/Models/DTOs/EmailTemplateDTO.cs:5]: "Dto" as postfix, not "DTO"

**Review by sulton-max — CHANGES_REQUESTED** [review]

**Review by botirjonabdumannonovv — APPROVED** [review]

**Review by botirjonabdumannonovv — APPROVED** [review]

**Review by Xolmuratov02 — DISMISSED** [review]

**Review by botirjonabdumannonovv — DISMISSED** [review]

File: src/Training.AirBnb.Clone.Backend/AirBnB.Api/Mappers/UserDto.cs [hunk 18]
```diff
@@ -0,0 +1,6 @@
+﻿namespace AirBnB.Api.Profiles
+{
+    internal class UserDto
+    {
```
[the remark this answers is not in the document]
Comment (sulton-max) [reply src/Training.AirBnb.Clone.Backend/AirBnB.Api/Mappers/UserDto.cs:4]: ha public qilish kerak

File: src/Training.AirBnb.Clone.Backend/AirBnB.Infrastructure/Common/Validators/EmailTemplateValidator .cs [hunk 19]
```diff
@@ -0,0 +1,24 @@
...
+            .MinimumLength(10)
+            .MaximumLength(129_536);
+
+        RuleFor(template => template.Type)
+            .Equal(NotificationType.Email);
+
+        RuleFor(emailTemplate => emailTemplate.Subject)
+            .NotEmpty()
+            .MinimumLength(10)
+            .MaximumLength(129_536);
```
Comment (sulton-max) [inline src/Training.AirBnb.Clone.Backend/AirBnB.Infrastructure/Common/Validators/EmailTemplateValidator .cs:22]: Max subject length must be 256

**Review by sulton-max — CHANGES_REQUESTED** [review]

File: src/Training.AirBnb.Clone.Backend/AirBnB.Api/Configurations/HostConfiguration.Extensions.cs [hunk 20]
```diff
@@ -1,3 +1,12 @@
...
+using AirBnB.Api.Data;
+using AirBnB.Application.Common.Identity.Services;
+using AirBnB.Application.Common.Settings;
+using AirBnB.Infrastructure.Common.Caching;
+using AirBnB.Infrastructure.Common.Identity.Services;
+using AirBnB.Infrastructure.Common.Settings;
+using AirBnB.Persistence.Caching.Brokers;
+
 using AirBnB.Persistence.DataContexts;
 using AirBnB.Persistence.Repositories.Interfaces;
```
Comment (botirjonabdumannonovv) [inline src/Training.AirBnb.Clone.Backend/AirBnB.Api/Configurations/HostConfiguration.Extensions.cs:11]: remove unnecessary row

**Review by botirjonabdumannonovv — DISMISSED** [review]

**Review by Xolmuratov02 — DISMISSED** [review]

**Review by sulton-max — DISMISSED** [review]

**Review by Xolmuratov02 — DISMISSED** [review]

**Pull request closed without merging**